src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+        return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+    }
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    req = ucp_stream_recv_request_get(ep->worker, buffer, *count, datatype, cb);
+    if (ucs_unlikely(req == NULL)) {
+        req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ep_stream = ep->ext.stream;
+
+    /* if there are already posted requests, need to queue this one */
+    if (!ucs_queue_is_empty(&ep_stream->reqs)) {
+        ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
+        goto out;
+    }
+
+    /* OK, lets obtain all arrived data which matches the recv size */
+    while (req->recv.state.offset < req->recv.length) {
+        rdesc = ucp_stream_recv_data_nb_internal(ep_stream);
+        if (rdesc == NULL) {
+            /* No data any more. Enqueue empty request or complete if some
+             * data is filled */
+            if (req->recv.state.offset == 0) {
+                ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
+            } else {
+                req = ucp_stream_recv_request_put(req, count, UCS_OK);
+            }
+            goto out;
+        }
+
+        status = ucp_stream_process_rdesc(rdesc, ep_stream, req);
+        if (ucs_unlikely(status != UCS_OK)) {
+            req = ucp_stream_recv_request_put(req, count, status);
+            goto out;
+        }
+    }
+
+    ucs_assert(req->recv.state.offset == req->recv.length);
+    req = ucp_stream_recv_request_put(req, count, UCS_OK);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+    return UCS_PTR_IS_PTR(req) ? (req + 1) : req;
+}
+
+static UCS_F_ALWAYS_INLINE ucp_recv_desc_t *
+ucp_stream_am_rdesc_get(ucp_worker_t *worker, void *data, size_t length,
+                        unsigned am_flags)
+{
+    ucp_recv_desc_t *rdesc;
 
     if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
         /* slowpath */
         rdesc        = (ucp_recv_desc_t *)data - 1;
         rdesc->flags = UCP_RECV_DESC_FLAG_UCT_DESC;
-        status       = UCS_INPROGRESS;
     } else {
-        status = UCS_OK;
+        /* TODO: optimize if there are posted recvs */
         rdesc = (ucp_recv_desc_t*)ucs_mpool_get_inline(&worker->am_mp);
         if (ucs_likely(rdesc != NULL)) {
             rdesc->flags = 0;
             memcpy(rdesc + 1, data, length);
         } else {
-            ucs_error("ucp recv descriptor is not allocated");
-            goto out;
+            ucs_fatal("ucp recv descriptor is not allocated");
         }
     }
 
-    rdesc->length  = recv_len;
-    rdesc->hdr_len = hdr_len;
-    hdr            = data;
+    rdesc->length  = length - sizeof(ucp_stream_am_hdr_t);
+    rdesc->hdr_len = sizeof(ucp_stream_am_hdr_t);
+
+    return rdesc;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_stream_ep_enqueue(ucp_ep_ext_stream_t *ep, ucp_worker_h worker)
+{
+    ucs_list_add_tail(&worker->stream_eps, &ep->list);
```

Comment:
need assert it's not already queued.
can also add helper function "uco_stream_ep_dequeue" and use it from ucp_stream_recv_data_nb_internal